libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,646 @@
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __advance_fwd(difference_type __offset, difference_type __steps) {
+    using __underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, __base_iter>>;
+    if constexpr (_Idx == sizeof...(_Views) - 1) {
+      std::get<_Idx>(__it_) += static_cast<__underlying_diff_type>(__steps);
+    } else {
+      difference_type __n_size = ranges::size(std::get<_Idx>(__parent_->__views_));
+      if (__offset + __steps < __n_size) {
+        std::get<_Idx>(__it_) += static_cast<__underlying_diff_type>(__steps);
+      } else {
+        __it_.template emplace<_Idx + 1>(ranges::begin(std::get<_Idx + 1>(__parent_->__views_)));
+        __advance_fwd<_Idx + 1>(0, __offset + __steps - __n_size);
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __advance_bwd(difference_type __offset, difference_type __steps) {
+    using __underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, __base_iter>>;
+    if constexpr (_Idx == 0) {
+      std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+    } else {
+      if (__offset >= __steps) {
+        std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+      } else {
+        auto __prev_size = ranges::__distance(std::get<_Idx - 1>(__parent_->__views_));
+        __it_.template emplace<_Idx - 1>(ranges::begin(std::get<_Idx - 1>(__parent_->__views_)) + __prev_size);
+        __advance_bwd<_Idx - 1>(__prev_size, __steps - __offset);
+      }
+    }
+  }
+
+  template <size_t... _Is, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func, index_sequence<_Is...>) {
+    ((__index == _Is ? (__func(integral_constant<size_t, _Is>{}), 0) : 0), ...);
+  }
+
+  template <size_t _Idx, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func) {
+    __apply_at_index(__index, std::forward<_Func>(__func), make_index_sequence<_Idx>{});
+  }
+
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI explicit constexpr __iterator(__maybe_const<_Const, concat_view>* __parent, _Args&&... __args)
+    requires constructible_from<__base_iter, _Args&&...>
+      : __it_(std::forward<_Args>(__args)...), __parent_(__parent) {}
+
+public:
+  _LIBCPP_HIDE_FROM_ABI __iterator() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator(__iterator<!_Const> __i)
+    requires _Const && (convertible_to<iterator_t<_Views>, iterator_t<const _Views>> && ...)
+      : __it_([&__src = __i.it_]<size_t... _Indices>(size_t __idx, index_sequence<_Indices...>) -> __base_iter {
+          using __src_lref          = decltype((__src));
+          using __construction_fptr = __base_iter (*)(__src_lref);
+          static constexpr __construction_fptr __vtable[]{[](__src_lref __src_var) -> __base_iter {
+            return __base_iter(in_place_index<_Indices>, std::get<_Indices>(std::move(__src_var)));
```

Comment:
I _guess_ `__unchecked_get` is desired because [LWG4079](https://cplusplus.github.io/LWG/issue4079) made the valueless case UB.

```suggestion
            return __base_iter(in_place_index<_Indices>, std::__unchecked_get<_Indices>(std::move(__src_var)));
```

But I'm now unsure which should be we should do in this PR, should we
- use `std::get` which possibly throws an exception, or
- use `__unchecked_get` only, which possibly raises (intended) UB, or
- use `__unchecked_get` together with some hardening checks?

CC @ldionne.

Reply:
Since LWG4079 makes this a precondition, it means UB if it's not satisfied. So I would use `__unchecked_get`, but I would indeed definitely add a hardening check along with a test so that we can catch it.

Reply:
This also makes me think that LWG4079 should be addressed (and marked as such) in this PR?

Reply:
Hi, whats an example of a hardening check I can follow to add?

Reply:
I'm not familiar to hardening. I _think_ we can include `<__assert>` in `<__ranges/concat_view.h>` and add asserting like

```C++
_LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(
    !__src.valueless_by_exception(), "Trying to convert from a valueless iterator of concat_view.");
```

before calling the function pointer.

Reply:
@changkhothuychung Please make sure to mark LWG4079 as implemented, closed and add tests.

Reply:
@Zingam I have look into how to add tests for this, just curious if its possible to do so. `valueless_by_exception ` is a member function of variant it seems. we have base_iter is a private member of class iterator. So im wondering how do we access this member to test?

Reply:
It seems that we can use `throw_on_move_iterator` in `libcxx/test/support/test_iterators.h` to test the valueless-by-exception case.